go/consensus/tendermint: Use MKVS for storing application state

kostko opened the pull request:
~~Based on #2659~~
~~Based on #2691~~
~~Based on #2715.~~
~~Based on #2775.~~
Fixes #1898 
Fixes #2710 
Fixes #2811 

Also makes some things nicer.

TODO
* [x] Support state rollback.
* [x] Tests for state rollback.
* [x] Do some stress tests.
* [x] Make set consensus parameters operations safer.

Yawning commented on go/consensus/tendermint/apps/staking/state/state.go:
```diff
@@ -76,66 +77,75 @@ type ImmutableState struct {
+	if err = cbor.Unmarshal(raw, &params); err != nil {
+		return nil, abci.UnavailableStateError(err)
+	}
 	return &params, err
```
Could return nil here for consistency.

kostko commented:
> How `setConesensusParameters` isn't still a massive foot + gun waiting to come back to cause problems escapes me.

Yeah that's just because I rebased on top of #2715 earlier so it has that exact change :-) I'll refactor that state to be more similar to other mux apps and changed our consensus parameter mutation operations to enforce that they can only be used in the proper context (e.g., in InitChain/BeginBlock/EndBlock).

kostko commented on go/consensus/tendermint/abci/state.go:
```diff
@@ -374,11 +386,31 @@ func (s *applicationState) doCleanup() {
+	return nil
+}
+
 func (s *applicationState) refreshConsensusParameters() error {
```
Yeah this is due to the rebase, will be removed.

kostko commented on go/consensus/tendermint/abci/state.go:
```diff
@@ -298,7 +306,11 @@ func (s *applicationState) OwnTxSigner() signature.PublicKey {
 func (s *applicationState) ConsensusParameters() (*consensusGenesis.Parameters, error) {
-	return &s.cachedConsensusParameters, nil
+	s.cachedConsensusParametersLock.RLock()
+	params := s.cachedConsensusParameters
```
Access from simulation mode which is async.

kostko replied:
This has now been removed, but the replacement is also under the (existing `blockLock`) lock.

kostko commented on go/consensus/tendermint/abci/mux.go:
```diff
@@ -740,6 +741,14 @@ func (mux *abciMux) DeliverTx(req types.RequestDeliverTx) types.ResponseDeliverT
 	if err := mux.executeTx(ctx, req.Tx); err != nil {
+		if IsUnavailableStateError(err) {
+			// Make sure to not commit any transactions which include results based on unavailable
+			// and/or corrupted state -- doing so can further corrupt state.
```
We wrap all state access and unmarshalling failures. Not sure if there are other cases that should be handled, but those could also use the special error wrapper.

kostko commented on go/consensus/tendermint/abci/prune.go:
```diff
@@ -60,111 +60,91 @@ type PruneConfig struct {
 
 type statePrunerInitializer interface {
-	Initialize(latestVersion int64) error
+	Initialize(latestRound uint64) error
```
Because our storage interface has a concept of rounds instead of versions (as it was previously only used for our runtime state). It is the same thing though.

kostko replied:
Renamed it back.

kostko commented on go/consensus/tendermint/abci/prune.go:
```diff
@@ -60,111 +60,91 @@ type PruneConfig struct {
-		"latest_version", latestVersion,
-		"start_version", p.eldestRetained,
+		"latest_version", latestRound,
+		"start_version", p.earliestRound,
```
Renamed it back.

kostko commented on go/consensus/tendermint/abci/context_test.go:
```diff
@@ -0,0 +1,82 @@
+	require.NoError(err, "Remove")
+
+	// Make sure updates didn't leak.
+	value, err = tree.Get(ctx, []byte("key"))
```
> is this a new behavior, that you can put a "State" result in your pocket and have access to a snapshot from the checkpoint?

So while you can retain a reference to the previous state, but if you modify it, these updates may leak to the checkpoint, so you shouldn't do that. We could try to prevent it by somehow invalidating all previous references but I'm not sure if it is worth the trouble.

> And you could update that outer tree while in the middle of a transaction.

As mentioned above, any updates to the outer tree may leak to the pending tree.

kostko replied:
I've added an explicit comment to `StartCheckpoint` mentioning previous references to `State`.

kostko commented:
The new way of updating consensus parameters is now in.

kostko commented on go/consensus/tendermint/abci/errors.go:
```diff
@@ -0,0 +1,32 @@
+// IsUnavailableStateError returns true if any error in err's chain is an unavailable state error.
+func IsUnavailableStateError(err error) bool {
+	var e *errorUnavailableState
+	return errors.As(err, &e)
```
Yeah, but `errors.Is` would not work because it checks for equality (e.g., all fields must be equal) while `errors.As` only requires assignment. Of course we could also implement the `Is(err error) bool` method, I'll probably do that to make it more explicit.

pro-wh replied:
oh good point

kostko commented on go/consensus/tendermint/abci/prune_test.go:
```diff
@@ -1,36 +1,77 @@
-		_, ver, err := tree.SaveVersion()
-		require.NoError(t, err, "SaveVersion: %d", i)
-		require.Equal(t, i, ver, "incorrect version on save")
+	require := require.New(t)
```
Sorry, I needed to move some files to `../api` which may mess with your bookmark.

kostko commented on go/consensus/tendermint/abci/context_test.go:
```diff
@@ -0,0 +1,82 @@
+	require.Nil(value, "updates should not leak outside checkpoint")
+
+	// Commit checkpoint.
+	cp.Commit()
```
Added.

kostko commented on go/consensus/tendermint/api/errors_test.go:
```diff
@@ -0,0 +1,24 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestErrors(t *testing.T) {
```
Good idea, added.

kostko commented on go/consensus/tendermint/api/errors.go:
```diff
@@ -0,0 +1,36 @@
+
+// UnavailableStateError wraps an error in an unavailable state error.
+func UnavailableStateError(err error) error {
+	if err == nil {
```
Thanks, fixed.

pro-wh commented on go/consensus/tendermint/api/context.go:
```diff
@@ -94,30 +85,56 @@ func FromCtx(ctx context.Context) *Context {
+		blockCtx:      blockCtx,
+		logger:        logging.GetLogger("consensus/tendermint/abci").With("mode", mode),
+	}
+	c.Context = context.WithValue(ctx, contextKey{}, c)
```
is this allowed inside the literal?

oh unless there's some logical separation to justify putting it here

kostko replied:
Oh you mean if we could move initializing the `Context` inside the literal that constructs `c`? Yeah the problem is that it requires a reference to `c` which doesn't exist yet at that point :-)

pro-wh commented on go/consensus/tendermint/abci/state.go:
```diff
@@ -114,101 +62,85 @@ type applicationState struct { // nolint: maligned
-// to view comitted data, and transactions to build the next version.
-func (s *applicationState) DeliverTxTree() *iavl.MutableTree {
-	return s.deliverTxTree
+	return s.blockParams
```
we unlock and return a pointer. so the rest of the codebase has to be respectful of the deep immutability

pro-wh commented on go/consensus/tendermint/abci/state.go:
```diff
@@ -331,71 +231,81 @@ func (s *applicationState) afterHaltEpoch(ctx *Context) bool {
+	s.blockLock.Lock()
+	defer s.blockLock.Unlock()
+
+	_, stateRootHash, err := s.deliverTxTree.Commit(s.ctx, s.stateRoot.Namespace, s.stateRoot.Round+1)
```
What's this new round thing that we use instead of height?

kostko replied:
It's the same thing, the only problem is that our MKVS has this concept of rounds (they are actually versions) because it was initially used only for runtime storage, which has a concept of rounds from the roothash module.

Now that this is more generally used it would probably make sense to change all "rounds" in MKVS to "versions".

kostko replied:
I've filed #2734 to make this more general now that we use MKVS in different places.

pro-wh commented on go/consensus/tendermint/abci/state.go:
```diff
@@ -421,78 +331,140 @@ func (s *applicationState) metricsWorker() {
+		select {
+		case <-s.ctx.Done():
+			return
+		case r := <-s.prunerNotifyCh.Out():
```
(me) bookmark

kostko commented on go/consensus/tendermint/abci/mux.go:
```diff
@@ -769,10 +771,18 @@ func (mux *abciMux) CheckTx(req types.RequestCheckTx) types.ResponseCheckTx {
+			ctx.Logger().Error("unavailable and/or corrupted state detected during tx processing",
+				"err", err,
+			)
+			panic(err)
```
Yeah, unfortunately ABCI does not provide a way to signal such failures -- the only way is to panic. This will cause a consensus failure and Commit will not be called so everything will be discarded.

pro-wh replied:
oh consensus commit. thanks

kostko commented on go/consensus/tendermint/abci/prune_test.go:
```diff
@@ -1,36 +1,77 @@
+		var rootHash hash.Hash
+		_, rootHash, err = tree.Commit(ctx, common.Namespace{}, i)
+		require.NoError(err, "Commit")
+		err = ndb.Finalize(ctx, common.Namespace{}, i, []hash.Hash{rootHash})
```
In case of Tendermint we could merge them (as there is always only one root per round which is the final root) to reduce some extra db round-trips. But most of those round-trips will be to in-memory memtables so it shouldn't account for much while a special case would add some code.

The reason why these two are separate in our MKVS is because we support the runtime state updates use case where multiple compute nodes can commit multiple roots. But only one is eventually finalized.

pro-wh replied:
oh thanks for the explanation

kostko commented on go/consensus/tendermint/abci/prune_test.go:
```diff
@@ -1,36 +1,77 @@
-	}
+	earliestVersion, err = ndb.GetEarliestRound(ctx)
+	require.NoError(err, "GetEarliestRound")
+	require.EqualValues(8, earliestVersion, "earliest version should be correct")
```
Yeah this is unchanged from before, constructing the pruner performs pruning.

kostko commented on go/consensus/tendermint/abci/prune_test.go:
```diff
@@ -1,36 +1,77 @@
-	require.True(t, tree.VersionExists(11), "VersionExists(11)")
+	earliestVersion, err = ndb.GetEarliestRound(ctx)
+	require.NoError(err, "GetEarliestRound")
+	require.EqualValues(9, earliestVersion, "earliest version should be correct")
```
Hm that's true but looking at the previous pruning test this exact same thing happened:
```golang
pruner.Prune(11)
require.False(t, tree.VersionExists(8), "VersionExists(8), should be pruned")
require.True(t, tree.VersionExists(11), "VersionExists(11)")
```

So version 8 was pruned, while 9, 10 and 11 existed. I guess this was already an off-by-one bug in the previous pruner implementation. I'll fix it here.

kostko replied:
Oh, actually this is probably correct, `NumKept` only includes historical versions. The latest version is always kept.

pro-wh replied:
o ok

kostko commented on go/consensus/tendermint/abci/state.go:
```diff
@@ -421,78 +331,140 @@ func (s *applicationState) metricsWorker() {
-		if err = s.refreshConsensusParameters(); err != nil {
+	if latestRound > 0 {
+		if err = s.doCommitOrInitChainLocked(time.Time{}); err != nil {
 			return nil, fmt.Errorf("state: failed to refresh consensus parameters: %w", err)
```
Ooops, thanks, fixed.

kostko commented on go/consensus/tendermint/abci/state.go:
```diff
@@ -421,78 +331,140 @@ func (s *applicationState) metricsWorker() {
+		DB:               filepath.Join(baseDir, storageDB.DefaultFileName(cfg.StorageBackend)),
+		MaxCacheSize:     64 * 1024 * 1024, // TODO: Make this configurable.
+		DiscardWriteLogs: true,
+		NoFsync:          true, // This is safe as Tendermint will replay on crash.
```
So in case the kernel/machine crashes while performing a commit, any Badger transactions are rolled back. The last thing that a `Finalize` does is to run a transaction that updates the last finalized round. In case that transaction is either not executed or is rolled back due to WAL corruption, when the node starts up again, `GetLatestRound` will return the previous round.

This will cause Tendermint (assuming that Tendermint's WAL implementation is correct) to see that the application is behind and will replay the block that was not committed or partially committed. So all updates will happen again and finalize will be called again.

Assuming that all state updates are deterministic, this is safe. In the worst case (only if the updates are non-deterministic) there would be some leftovers in the database which would not be garbage collected (this could happen even if fsync is enabled).

pro-wh replied:
ok so NoFsync won't cause inconsistency

kostko commented on go/consensus/tendermint/abci/timer.go:
```diff
@@ -144,57 +149,65 @@ func (t *Timer) Reset(ctx *Context, duration time.Duration, data []byte) {
 	}
 
-	if _, removed := ctx.State().Remove(t.ID); !removed {
+	if existing, err := ctx.State().RemoveExisting(ctx, t.ID); existing == nil || err != nil {
```
Nice catch, added to the panic message.

kostko commented:
Last commit shows no performance degradation under pruning after a few days of runtime and over 350k processed blocks.

Storing all root metadata under the same key for each round/version does slow down specific unit tests which generate over a thousand roots in the same version (locally 6s, on CI over 30s for some reason).

kostko commented:
Split Badger MKVS changes into #2762.

pro-wh commented on go/consensus/tendermint/apps/registry/state/state.go:
```diff
@@ -552,123 +572,156 @@ func NewImmutableState(state abci.ApplicationState, version int64) (*ImmutableSt
-	s.tree.Remove(signedRuntimeKeyFmt.Encode(&id))
-	s.tree.Set(suspendedRuntimeKeyFmt.Encode(&id), raw)
-	return nil
+	return s.Tree.Insert(ctx, suspendedRuntimeKeyFmt.Encode(&id), data)
```
ought this one be wrapped in UnavailableStateError?

kostko replied:
Ah, yeah you're correct, it should be (also in other places). No shortcuts :P

pro-wh commented on go/consensus/tendermint/apps/registry/state/state.go:
```diff
@@ -552,123 +572,156 @@ func NewImmutableState(state abci.ApplicationState, version int64) (*ImmutableSt
-	s.tree.Remove(suspendedRuntimeKeyFmt.Encode(&id))
-	s.tree.Set(signedRuntimeKeyFmt.Encode(&id), raw)
-	return nil
+	return s.Tree.Insert(ctx, signedRuntimeKeyFmt.Encode(&id), data)
```
here too

pro-wh commented on go/consensus/tendermint/apps/roothash/genesis.go:
```diff
@@ -30,19 +33,24 @@ func (app *rootHashApplication) InitChain(ctx *abci.Context, request types.Reque
 		)
-		app.onNewRuntime(ctx, v, &st)
+		if err := app.onNewRuntime(ctx, v, &st); err != nil {
+			return fmt.Errorf("failed to initialize runtime state: %w", err)
```
could add the runtime ID in the context

kostko replied:
Added.

pro-wh commented on go/consensus/tendermint/apps/roothash/genesis.go:
```diff
@@ -30,19 +33,24 @@ func (app *rootHashApplication) InitChain(ctx *abci.Context, request types.Reque
-	runtimes := rq.state.Runtimes()
+	runtimes, err := rq.state.Runtimes(ctx)
+	if err != nil {
+		return nil, err
```
in most places you've been wrapping these with context info

kostko replied:
Added.

pro-wh commented on go/consensus/tendermint/apps/roothash/roothash.go:
```diff
@@ -358,15 +355,16 @@ func (app *rootHashApplication) ForeignExecuteTx(ctx *abci.Context, other abci.A
 						"runtime", rt.ID,
 					)
 
-					ensureGenesis()
```
:information_source: turns out the genesis arg of onNewRuntime is only used in InitChian

pro-wh commented on go/consensus/tendermint/apps/roothash/roothash.go:
```diff
@@ -476,14 +484,18 @@ func (app *rootHashApplication) FireTimer(ctx *abci.Context, timer *abci.Timer)
 
-	defer state.SetRuntimeState(rtState)
+	defer func() {
+		if err2 := state.SetRuntimeState(ctx, rtState); err2 != nil {
```
here too, why do we insist on saving this in case of error? I think whatever calls into this can't recover from anything

kostko replied:
Good point, this was actually strange even in the previous version.

kostko replied:
Changed it to not use defer at all.

pro-wh commented on go/consensus/tendermint/apps/roothash/roothash.go:
```diff
@@ -682,16 +694,15 @@ func (app *rootHashApplication) postProcessFinalizedBlock(ctx *abci.Context, rtS
 		}
 	}
 
+	checkpoint.Commit()
```
different pattern from elsewhere, which used an `ok` flag and a branch in the `defer` func to either close or commit. this one relies on `Commit(); Close()` being legal

kostko replied:
Ah that's a good point, I think this is actually a nicer pattern.

> this one relies on Commit(); Close() being legal

Which is.

kostko replied:
Changed it so this pattern is used everywhere.

pro-wh replied:
ok

kostko commented on go/consensus/tendermint/abci/prune.go:
```diff
@@ -60,111 +59,99 @@ type PruneConfig struct {
+			p.logger.Debug("Prune: skipping non-earliest version",
+				"version", i,
+			)
+			continue
```
In case the given version is earlier than the earliest.

I think the only time this happens is when the pruner is initialized with an empty version which is zero. But then a new higher version is created (as the first version can be arbitrary), so this would keep on failing.

Actually now that I've explained it it seems that the pruner should just detect this empty case and re-check what the earliest version is.

kostko commented on go/consensus/tendermint/abci/timer.go:
```diff
@@ -144,57 +149,65 @@ func (t *Timer) Reset(ctx *Context, duration time.Duration, data []byte) {
+
+		// Skip timers that are not for this application.
+		if app.ID() != ts.app {
+			continue
```
Uh, that's a good idea, the existing way seems stupid :man_facepalming:

kostko commented on go/consensus/tendermint/api/state.go:
```diff
@@ -0,0 +1,245 @@
+		// - If this request was made from an ABCI app and is for the current (future) height.
+		//
+		if abciCtx.IsInitChain() || version == abciCtx.BlockHeight()+1 {
+			return &ImmutableState{Tree: abciCtx.State()}, nil
```
It doesn't really, we assume that users actually respect that. We could introduce an `ImmutableKeyValueTree` interface which doesn't have the mutation operators to make this more explicit.

kostko replied:
Introduced an `ImmutableKeyValueTree` so this is now more explicit.

kostko commented on go/consensus/tendermint/api/state.go:
```diff
@@ -0,0 +1,245 @@
+		return nil, err
+	}
+	if len(roots) != 1 {
+		return nil, fmt.Errorf("state: corrupted state (%d): %+v", version, roots)
```
I think I already explained this -- so our MKVS supports multiple tree roots per version (height in this case). This is because for runtimes we store both IO roots and state roots and there can be multiple of them until the version is finalized (because until discrepancy detection happens, you don't know which root is correct). This is also the reason why `Commit` and `Finalize` are split.

Of course in the case where MKVS is used for consensus layer state there is always only one root per version so this looks strange. I was thinking we could do some optimizations for this case and also enforce it at the database level by having a `MaxRootsPerVersion` config option.

pro-wh replied:
ok. thanks for the explanation

kostko commented on go/consensus/tendermint/apps/beacon/state/state.go:
```diff
@@ -37,53 +37,58 @@ func NewImmutableState(state abci.ApplicationState, version int64) (*ImmutableSt
-	err := cbor.Unmarshal(raw, &params)
-	return &params, err
+	if err = cbor.Unmarshal(data, &params); err != nil {
+		return nil, abciAPI.UnavailableStateError(err)
```
Yeah the assumption here is that if unmarshal fails then it must be due to some kind of data corruption which may be non-deterministic. Of course it could also be deterministic in case of bugs that deterministically corrupt state, but you can't know.

pro-wh replied:
oh

kostko commented on go/consensus/tendermint/apps/beacon/state/state.go:
```diff
@@ -37,53 +37,58 @@ func NewImmutableState(state abci.ApplicationState, version int64) (*ImmutableSt
-	s.tree.Set(parametersKeyFmt.Encode(), cbor.Marshal(params))
+func (s *MutableState) SetConsensusParameters(ctx context.Context, params *beacon.ConsensusParameters) error {
+	// TODO: Enforce that we are in InitChain/BeginBlock/EndBlock.
+	err := s.Tree.Insert(ctx, parametersKeyFmt.Encode(), cbor.Marshal(params))
```
Should be fine since this is actually in `MutableState`. As mentioned above, we could use explicit interfaces to make this more clear?

kostko commented on go/consensus/tendermint/apps/keymanager/state/state.go:
```diff
@@ -31,47 +32,49 @@ func (st *ImmutableState) Statuses() ([]*api.Status, error) {
+		if !statusKeyFmt.Decode(it.Key()) {
+			break
+		}
+		rawVec = append(rawVec, it.Value())
```
Yes, the most that can happen is that the in-memory tree node is evicted from cache but even in that case the value slice is never reused.

pro-wh replied:
fine. it's an unwritten property though

kostko commented on go/consensus/tendermint/api/context.go:
```diff
@@ -253,34 +274,50 @@ func (c *Context) BlockContext() *BlockContext {
+// StartCheckpoint starts a new state checkpoint. Any further updates to the context's state will
+// be performed against the checkpoint and will only be committed in case of an explicit Commit.
+//
+// Any existing references to the underlying State() should not be mutated while the checkpoint is
```
Made it more clear, thanks.

kostko commented on go/consensus/tendermint/apps/epochtime_mock/epochtime_mock.go:
```diff
@@ -40,28 +40,30 @@ func (app *epochTimeMockApplication) Dependencies() []string {
 	}
-	defer state.clearFutureEpoch()
+	defer func() {
+		_ = state.clearFutureEpoch(ctx)
```
Changed this around, no defer needed.

kostko commented on go/consensus/tendermint/apps/registry/registry.go:
```diff
@@ -146,7 +146,9 @@ func (app *registryApplication) onRegistryEpochChanged(ctx *abci.Context, regist
 		}
-		defer stakeAcc.Commit()
+		defer func() {
+			err = stakeAcc.Commit()
```
Yeah, turned this around without defer.

kostko commented:
> immutability is probably the wrong word for it. it's more of a read only interface, with no consistency guarantee

So the current terminology comes from IAVL.

Why no consistency guarantee? If you create an ImmutableState, you create it at a specific height and that behaves like a snapshot. It is only if you create it from inside the ABCI app while processing a block where this snapshot is actually mutable (because it is not committed yet).

kostko commented on go/consensus/tendermint/apps/roothash/state/state.go:
```diff
@@ -37,82 +38,102 @@ type RuntimeState struct {
+	if err = cbor.Unmarshal(raw, &state); err != nil {
+		return nil, abciAPI.UnavailableStateError(err)
+	}
 	return &state, err
```
Ah this is a leftover, thanks.

kostko commented on go/consensus/tendermint/apps/scheduler/scheduler.go:
```diff
@@ -662,12 +664,24 @@ func publicKeyMapToSliceByStake(
-		iBal := stakeAcc.GetEscrowBalance(entities[i])
-		jBal := stakeAcc.GetEscrowBalance(entities[j])
-		return iBal.Cmp(&jBal) == 1 // Note: Not -1 to get a reversed sort.
+		iBal, err := stakeAcc.GetEscrowBalance(entities[i])
```
The stake accumulator caches everything, so this should be fine.

kostko commented on go/consensus/tendermint/api/errors.go:
```diff
@@ -0,0 +1,40 @@
+
+// UnavailableStateError wraps an error in an unavailable state error.
+func UnavailableStateError(err error) error {
+	if err == nil {
```
Hm good point, maybe we should change the function to also take context information?

kostko commented on go/consensus/tendermint/apps/roothash/transactions.go:
```diff
@@ -119,7 +121,11 @@ func (app *rootHashApplication) executorCommit(
 	}
-	defer state.SetRuntimeState(rtState)
+	defer func() {
+		if err2 := state.SetRuntimeState(ctx, rtState); err2 != nil {
```
I think one idea was that since commitments are independently verified, we should still add all that are valid, even if some are not. However, we currently fail on the first error with executor commitments, so this would only be applicable to merge commitments.

So on second thought I've now changed this to only commit runtime state in case all commitments were valid and there were no other errors.

kostko commented on go/consensus/tendermint/apps/scheduler/state/state.go:
```diff
@@ -33,163 +32,188 @@ var (
-	)
+		var c api.Committee
+		if err := cbor.Unmarshal(it.Value(), &c); err != nil {
+			return nil, abciAPI.UnavailableStateError(err)
```
Added.

kostko commented on go/consensus/tendermint/apps/scheduler/state/state.go:
```diff
@@ -33,163 +32,188 @@ var (
+
+			var c api.Committee
+			if err := cbor.Unmarshal(it.Value(), &c); err != nil {
+				return nil, abciAPI.UnavailableStateError(err)
```
Added.

kostko commented on go/consensus/tendermint/apps/registry/registry.go:
```diff
@@ -196,6 +197,12 @@ func (app *registryApplication) onRegistryEpochChanged(ctx *abci.Context, regist
 		}
 	}
 
+	if stakeAcc != nil {
```
Unified, thanks.

kostko commented:
> snapshots: hand out overlay objects instead of snapshot objects. could make it easier manage finer and finer grained rollbacks. could make it easier to manage non-commit at the top ABCI entry in case of unavailable state error

The only potential issue I see with this is that it would be easier to misuse `ctx.State()` by accident instead of using the correct overlay. You could have finer grained rollbacks with the current API if we allowed nesting checkpoints (we currently do not).

kostko commented on go/consensus/tendermint/apps/scheduler/scheduler.go:
```diff
@@ -456,8 +460,7 @@ func (app *schedulerApplication) electCommittee(
 		)
-		schedulerState.NewMutableState(ctx.State()).DropCommittee(kind, rt.ID)
-		return nil
+		return schedulerState.NewMutableState(ctx.State()).DropCommittee(ctx, kind, rt.ID)
```
Good idea, added.

kostko commented on go/consensus/tendermint/abci/state/state.go:
```diff
@@ -0,0 +1,79 @@
+
+// ImmutableState is an immutable consensus backend state wrapper.
+type ImmutableState struct {
+	is *api.ImmutableState
```
This is so that the underlying methods are not public as if we just embed all the raw tree accessors will be exposed.

pro-wh commented:
this reviewer interface is showing a bunch of commits that were rebased in #2775

kostko commented:
Here you go, rebased.

pro-wh commented:
thanks

pro-wh commented on go/roothash/api/commitment/pool.go:
```diff
@@ -173,7 +174,7 @@ func (p *Pool) addOpenExecutorCommitment(blk *block.Block, sv SignatureVerifier,
 	// Verify RAK-attestation.
 	if p.Runtime.TEEHardware != node.TEEHardwareInvalid {
-		n, err := nl.Node(id)
+		n, err := nl.Node(context.TODO(), id)
```
could encapsulate the context in the NodeLookup implementation. or if the NodeLookup needs to be long lived, could pass in context from caller

kostko replied:
Passing context from caller seems cleaner, added.

pro-wh commented:
> > immutability is probably the wrong word for it. it's more of a read only interface, with no consistency guarantee
> 
> So the current terminology comes from IAVL.
> 
> Why no consistency guarantee? If you create an ImmutableState, you create it at a specific height and that behaves like a snapshot. It is only if you create it from inside the ABCI app while processing a block where this snapshot is actually mutable (because it is not committed yet).

this is mostly for trees of the "latest" block-in-progress state. there can be ImmutableStates and MutableStates backed by the same mkvs store, which is overall mutable

pro-wh commented:
> > snapshots: hand out overlay objects instead of snapshot objects. could make it easier manage finer and finer grained rollbacks. could make it easier to manage non-commit at the top ABCI entry in case of unavailable state error
> 
> The only potential issue I see with this is that it would be easier to misuse `ctx.State()` by accident instead of using the correct overlay. You could have finer grained rollbacks with the current API if we allowed nesting checkpoints (we currently do not).

could pair with a context wrapping technique that makes subordinate contexts return the overlay in its `State()` implementation.

kostko commented on go/consensus/tendermint/apps/staking/query.go:
```diff
@@ -87,23 +69,23 @@ func (sq *stakingQuerier) Threshold(ctx context.Context, kind staking.ThresholdK
+	return sq.state.DebondingDelegationsFor(ctx, id)
 }
 
 func (sq *stakingQuerier) ConsensusParameters(ctx context.Context) (*staking.ConsensusParameters, error) {
```
Yeah it won't build otherwise anyway ;-)

kostko commented on go/consensus/tendermint/api/state.go:
```diff
@@ -0,0 +1,252 @@
+	switch len(roots) {
+	case 0:
+		// No roots for that state -- it may have been pruned.
+		return nil, consensus.ErrVersionNotFound
```
Nothing else changed, I just changed the error after the new query workload triggered this when pruning is enabled. This makes the error message more correct as a missing version is not state corruption.

pro-wh replied:
ok thanks

kostko commented on go/consensus/tendermint/apps/roothash/state/state.go:
```diff
@@ -37,82 +38,102 @@ type RuntimeState struct {
+// SetConsensusParameters sets roothash consensus parameters.
+func (s *MutableState) SetConsensusParameters(ctx context.Context, params *roothash.ConsensusParameters) error {
+	err := s.ms.Insert(ctx, parametersKeyFmt.Encode(), cbor.Marshal(params))
+	return abciAPI.UnavailableStateError(err)
```
It was your comment about the leftover on line 69 (returning err instead of nil), which should now be fixed.

pro-wh replied:
ah. github had detached that thread. thanks for the reminder

kostko commented on go/consensus/tendermint/apps/roothash/transactions.go:
```diff
@@ -140,20 +141,25 @@ func (app *rootHashApplication) executorCommit(
 
+	// Update runtime state.
+	if err = state.SetRuntimeState(ctx, rtState); err != nil {
+		return fmt.Errorf("failed to set runtime state: %w", err)
```
> how might adding commitments fail? like if one has an invalid signature, then the rest are also not stored? I guess that doesn't matter much in this case because they're all coming from the same transaction sender

Yeah invalid signature, not based on correct block, not in committee, etc. So currently compute nodes only send a single commitment (their own) so this is not a problem in practice. If we had commit aggregators we would probably want to just skip over invalid commits instead of returning an error.

pro-wh replied:
ok yeah. aggregators ought to check too, so things shouldn't get too crazy even if we forget about this

kostko commented on go/consensus/tendermint/apps/staking/state/gas.go:
```diff
@@ -89,22 +92,26 @@ func AuthenticateAndPayFees(
 	state := NewMutableState(ctx.State())
-	state.SetLastBlockFees(&fees)
+	if err := state.SetLastBlockFees(ctx, &fees); err != nil {
+		panic(err)
```
Good point, changed to return an error instead.

kostko commented on go/consensus/tendermint/apps/staking/staking.go:
```diff
@@ -156,51 +155,69 @@ func (app *stakingApplication) EndBlock(ctx *abci.Context, request types.Request
 		if !e.DelegatorID.Equal(e.EscrowID) {
-			state.SetAccount(e.EscrowID, escrow)
+			if err = state.SetAccount(ctx, e.EscrowID, escrow); err != nil {
+				return fmt.Errorf("failed to set account: %w", err)
```
Added.
